recipes/openai_textcat_multilabel.py

```diff
@@ -0,0 +1,527 @@
+    ) -> str:
+        """Generate a prompt for multilabel text categorization."""
+        prompt = self.prompt_template.render(
+            text=text, labels=labels, exclusive=self.exclusive_classes
+        )
+        return prompt
+
+    def _get_textcat_response(self, prompts: List[str], delay: int = 0) -> List[str]:
+        headers = {
+            "Authorization": f"Bearer {self.openai_api_key}",
+            "OpenAI-Organization": self.openai_api_org,
+            "Content-Type": "application/json",
+        }
+        time.sleep(delay)
+        r = _retry429(
+            lambda: client.post(
+                "https://api.openai.com/v1/completions",
+                headers=headers,
+                json={
+                    "model": self.model,
+                    "prompt": prompts,
+                    "temperature": self.openai_temperature,
+                    "max_tokens": self.openai_max_tokens,
+                },
+            ),
+            n=self.openai_n,
+            timeout_s=self.openai_timeout_s,
+        )
+        r.raise_for_status()
+        responses = r.json()
+        return [responses["choices"][i]["text"] for i in range(len(prompts))]
+
+    def _parse_response(self, text: str) -> Dict[str, str]:
+        """Interpret OpenAI's TextCat response. It's supposed to be
+        a list of lines, with each line having the form:
+
+        answer: <accept/reject for binary or the label for multilabel>
+        reason: <string>
+
+        However, there's no guarantee that the model will give
+        us well-formed output. It could say anything, it's an LM.
+        So we need to be robust.
+        """
+        output = {}
+        for line in text.strip().split("\n"):
+            if line and ":" in line:
+                key, value = line.split(":", 1)
+                output[key.strip().lower()] = value.strip()
+        return output
+
+
+@prodigy.recipe(
+    # fmt: off
+    "textcat_multilabel.openai.correct",
+    input_path=("Path to jsonl data to annotate", "positional", None, Path),
+    labels=("Labels (comma delimited)", "positional", None, lambda s: s.split(",")),
+    lang=("Language to initialize spaCy model", "option", "l", str),
+    model=("GPT-3 model to use for completion", "option", "m", str),
+    batch_size=("Batch size to send to OpenAI API", "option", "b", int),
+    segment=("Split sentences", "flag", "S", bool),
+    examples_path=("Examples file to help define the task", "option", "e", Path),
+    prompt_path=("Path to jinja2 prompt template", "option", "p", Path),
+    max_examples=("Max examples to include in prompt", "option", "n", int),
+    exclusive_classes=("Make the classification task exclusive", "flag", "E", bool),
+    verbose=("Print extra information to terminal", "flag", "v", bool),
+    # fmt: on
+)
+def textcat_multilabel_openai_correct(
+    dataset: str,
+    input_path: Path,
+    labels: List[str],
```

Comment:
In the case of a single label, the interface looks like this:

![CleanShot 2023-01-17 at 13 27 42](https://user-images.githubusercontent.com/1019791/212899110-83a669c0-b329-40cb-9d52-7876c478adf6.png)

Wouldn't it be better to generate a binary interface for it instead? Something like: 

![CleanShot 2023-01-17 at 13 28 49](https://user-images.githubusercontent.com/1019791/212899320-eb2bad5e-ab4b-42e2-b5f2-b945fbc10635.png)

Textcat manual automatically takes care of this, maybe we can do that here as well?

Reply:
I did a bit of refactor that combines both of them. Will update the PR description on the changes. But yes, this should be automatic now